docs: context gateway (#1789)
shopware/docs: guides/plugins/apps/gateways/context/context-gateway.md

# Context Gateway

::: danger
**Security and privacy**

With the context gateway, Shopware allows your app to manipulate the customer context, which includes sensitive information like customer addresses, payment methods, and more.
It is your responsibility to ensure that the commands are valid and do not compromise the security or privacy of customers.

As this is a powerful feature, you should only use it if you are sure that your app server is secure and the commands you send are safe.

:::

## Context

> As of Shopware version 6.7.1.0, the Context Gateway was introduced.
> 
> The Context Gateway is a powerful feature that allows apps to securely interact with the customer context, based on the current cart and sales channel and make informed decisions on the app server.
> The app system benefits from this solution, enabling app developers to manipulate customer contexts and provide a more tailored shopping experience.
> 
> It is designed as to being the connection point between your apps Javascript and your app server.

Reviewer: ```suggestion
As of Shopware version 6.7.1.0, the Context Gateway has been introduced.

The Context Gateway is a powerful feature that enables apps to securely access and interact with the customer context — based on the current cart and sales channel — allowing for more informed decision-making on the app server. This enhancement empowers app developers to dynamically tailor the shopping experience by manipulating the customer context.

It serves as the bridge between your app’s JavaScript and your app server.
```

## Prerequisites

You should be familiar with the concept of Apps, their registration flow as well as signing and verifying requests and responses between Shopware and the App backend server.

<PageRef page="../../app-base-guide.md" title="App base guide" />

> Your app server must be also accessible for the Shopware server.

Reviewer: ```suggestion
Your app server must be also accessible to the Shopware server.
```

> You can use a tunneling service like [ngrok](https://ngrok.com/) for development.

Reviewer 2: Why would you need a tunneling service? Can't you just have Shopware on one port and the app server on another? For development purposes, I don't see the reason.

## Manifest configuration

To indicate to Shopware that your app uses the context gateway, you must provide a `context` property inside a `gateways` parent property of your app's `manifest.xml`.

Below, you can see an example definition of a working checkout gateway configuration.

::: code-group

```xml [manifest.xml]
<?xml version="1.0" encoding="UTF-8"?>
<manifest>
    <!-- ... -->

    <gateways>
        <!-- ... -->
        <context>https://my-app.server.com/context/gateway</context>
    </gateways>
</manifest>
```

:::

After successful installation of your app, the context gateway is ready to be called by Shopware.

## Context gateway endpoint

> Your JavaScript integration is responsible for calling the context gateway endpoint.

Reviewer 2: I think, this needs clarification. I would not make this JS specific. I would describe, that an endpoint on the Shopware server needs to be called (either Store-API or Storefront), which then in turn calls the app server with the data that is passed. This ensures secured communication between Shopware and the App Server.

The app server will receive the current `SalesChannelContext`, `Cart`, and any custom data you provide as part of the payload.

::: warning
**Connection timeouts**

The Shopware shop will wait for a response for 5 seconds.
Be sure that your context gateway implementation on your app server responds in time, otherwise Shopware will time out and drop the connection.
:::

### JavaScript `ContextGatewayClient`

> Your implementation must call the `frontend.gateway.context` storefront endpoint, which is automatically registered by Shopware beforehand.

Reviewer 2: This is Twig Storefront specific and should be marked as such.

> You can send any data in the requests body, which will be forwarded to your app server.

Reviewer: ```suggestion
You can send any data in the body of the request, which will be forwarded to your app server.
```

> Luckily, Shopware provides you with a `ContextGatewayClient` service, which you can use to ease the communication with the context gateway.

Reviewer: ```suggestion
Shopware provides a `ContextGatewayClient` service, which can be used to ease the communication with the context gateway.
```

            $commands->add(new ChangeCurrencyCommand('GBP'));
            $commands->add(new ChangeLanguageCommand('en-GB'));
        }

        $response = GatewayResponse::createContextGatewayResponse($commands);

        return $this->httpFoundationFactory->createResponse($response);
    }
}
```

</Tab>

</Tabs>

### Command Validation

Shopware performs basic validation on the commands returned by your app server to ensure they are reasonable to execute.

The following checks are enforced:

- The command must be recognized as valid, e.g. <nobr>`context_change-currency`</nobr>. See the full list of available [commands](./command-reference.md#available-commands).
- The payload must be valid for the respective command type.
- Only **one command per type** is allowed. For example, you cannot include two <nobr>`context_change-currency`</nobr> commands in a single response.
- A maximum of **one <nobr>`context_register-customer`</nobr> or <nobr>`context_login-customer`</nobr>** command is allowed per response.

## Event

Plugins can listen to the `Shopware\Core\Framework\Gateway\Context\Command\Event\ContextGatewayCommandsCollectedEvent`.

> This event is dispatched after all commands have been collected from the app server and allow plugins to modify or add commands based on the same payload the app received. 

Reviewer 3: ```suggestion
This event is dispatched after all commands have been collected from the app server and allow plugins to modify or add commands based on the same payload the app received.
```

use Symfony\Component\Routing\Attribute\Route;

#[Route('/api/gateway', name: 'api.gateway.')]
class GatewayController extends AbstractController
{
    public function __construct(
        private readonly HttpFoundationFactoryInterface $httpFoundationFactory
    ) {
    }

    #[Route('/context', name: 'context', methods: ['POST'])]
    public function context(ContextGatewayAction $action): Response
    {
        /** @var Collection<ContextGatewayCommand> $commands */
        $commands = new Collection();

        if ($action->getSalesChannelContext()->getCurrency()->getIsoCode() !== 'GBP') {
            $commands->add(new ChangeCurrencyCommand('GBP'));
            $commands->add(new ChangeLanguageCommand('en-GB'));
        }

        $response = GatewayResponse::createContextGatewayResponse($commands);

        return $this->httpFoundationFactory->createResponse($response);
    }
}
```

</Tab>

</Tabs>

### Command Validation

Shopware performs basic validation on the commands returned by your app server to ensure they are reasonable to execute.

The following checks are enforced:

- The command must be recognized as valid, e.g. <nobr>`context_change-currency`</nobr>.

>   See the [full list of available commands](./command-reference.md#available-commands).

Reviewer 3: ```suggestion
- The command must be recognized as valid, e.g. <nobr>`context_change-currency`</nobr>. See the full list of available [commands](./command-reference.md#available-commands).
```
